core/state/metadata.go

```diff
@@ -0,0 +1,39 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package state
+
+// BalanceChangeReason is used to indicate the reason for a balance change, useful
+// for tracing and reporting.
+type BalanceChangeReason byte
+
+const (
+	BalanceChangeUnspecified BalanceChangeReason = iota
+	BalanceChangeRewardMineUncle
+	BalanceChangeRewardMineBlock
+	BalanceChangeDaoRefundContract
+	BalanceChangeDaoAdjustBalance
+	BalanceChangeTransfer
+	BalanceChangeGenesisBalance
+	BalanceChangeGasBuy
+	BalanceChangeRewardTransactionFee
+	BalanceChangeGasRefund
+	BalanceChangeTouchAccount
+	BalanceChangeSuicideRefund
+	BalanceChangeSuicideWithdraw
```

Comment:
```suggestion
	BalanceChangeSelfdestructDeduction
	BalanceChangeSelfdestructAddition
```
or something

Reply:
Aslo, the actual selfdestruct burn happens without SetBalance - it does a raw `data.Balance` setting
```golang
func (s *StateDB) SelfDestruct(addr common.Address) {
	stateObject := s.getStateObject(addr)
	if stateObject == nil {
		return
	}
	s.journal.append(selfDestructChange{
		account:     &addr,
		prev:        stateObject.selfDestructed,
		prevbalance: new(big.Int).Set(stateObject.Balance()),
	})
	stateObject.markSelfdestructed()
	stateObject.data.Balance = new(big.Int)
}
```

Reply:
I would like you to change these a bit, and also document them properly. Perhaps split them up into `Increase`, `Decrease` or `Change` (both/either). 

It could look like this: 
```
	// BalanceChangeTransfer is a change due to either incoming or outgoing value-transferring call
	BalanceChangeTransfer
	
	// BalanceIncreaseGenesis is an increase of balance due to genesis allocation
	BalanceIncreaseGenesis

	// BalanceChangeGasBuy is a deduction of balance due to purchasing gas.
	BalanceDecreaseGasPurchase

	// BalanceChangeRewardTransactionFee is an increase of balance due to transaction fees.
	BalanceIncreaseRewardTransactionFee

	// BalanceChangeGasRefund is an increase of balance due to refund of gas.
	BalanceIncreaseGasRefund

	// BalanceIncreaseSelfdestruct is an increase of balance for the beneficiary of a selfdestruct.
	BalanceIncreaseSelfdestruct

	//BalanceDecreaseSelfdestruct is a decrease of balance due to selfdestruct.
	BalanceDecreaseSelfdestruct

	// BalanceIncreaseWithdrawal is an increase of balance due to withdrawal 
	BalanceIncreaseWithdrawal
```
I removed unused ones, but the ones at the top I just didn't bother fixing so that's a todo

Reply:
Applied these changes with some adaptation, thanks!